packages/core/src/browser/preferences/preference-service.ts

```diff
@@ -16,25 +16,51 @@
 
 // tslint:disable:no-any
 
-import { JSONExt } from '@phosphor/coreutils';
 import { injectable, inject, postConstruct } from 'inversify';
 import { FrontendApplicationContribution } from '../../browser';
 import { Event, Emitter, DisposableCollection, Disposable, deepFreeze } from '../../common';
 import { Deferred } from '../../common/promise-util';
 import { PreferenceProvider } from './preference-provider';
 import { PreferenceSchemaProvider } from './preference-contribution';
+import URI from '../../common/uri';
 
 export enum PreferenceScope {
-    User,
-    Workspace
+    DEFAULT,
+    USER,
+    WORKSPACE,
+    FOLDER
 }
 
-export interface PreferenceChangedEvent {
-    changes: PreferenceChange[]
+export namespace PreferenceScope {
+    export function getScopes(): PreferenceScope[] {
+        return Object.keys(PreferenceScope)
+            .filter(k => typeof PreferenceScope[k as any] === 'string')
+            .map(v => <PreferenceScope>Number(v));
+    }
+
+    export function getReversedScopes(): PreferenceScope[] {
+        return getScopes().reverse();
+    }
+
+    export function getScopeNames(scope?: PreferenceScope): string[] {
+        const names: string[] = [];
+        const allNames = Object.keys(PreferenceScope)
+            .filter(k => typeof PreferenceScope[k as any] === 'number');
+        if (scope) {
+            for (const name of allNames) {
+                if ((<any>PreferenceScope)[name] <= scope) {
+                    names.push(name);
+                }
+            }
+        }
+        return names;
+    }
 }
 
 export interface PreferenceChange {
     readonly preferenceName: string;
+    readonly scope: PreferenceScope
+    readonly domain: string[];
```

Comment:
it is an implementation detail, end users should not care, they already have `scope` and a function `affects` to check whether their resource is affected by a change

We should hide `domain` as implementation details.

Reply:
i reverted my changes to `PreferenceChange`.